lints/etsi/lint_qcstatem_psd2_policy_mandatory.go

```diff
@@ -0,0 +1,80 @@
+package etsi
+
+/*
+ * ZLint Copyright 2020 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"encoding/asn1"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/lint"
+	"github.com/zmap/zlint/util"
+)
+
+type qcStatemPsd2PolicyMandatory struct{}
+
+func (l *qcStatemPsd2PolicyMandatory) Initialize() error {
+	return nil
+}
+
+func (l *qcStatemPsd2PolicyMandatory) CheckApplies(c *x509.Certificate) bool {
+	isPresent := util.HasCertAnyEtsiQcStatement(c)
+	return isPresent
+}
+
+func policyAndTypeAreConsistent(c *x509.Certificate, policy asn1.ObjectIdentifier, policyStr string, qcType asn1.ObjectIdentifier, typeStr string, mustHaveQcStmtFour bool) string {
+
+	if util.HasCertPolicy(c, policy) && !util.HasCertEtsiQcType(c, qcType) {
+		return "EU Qualified Certificate has policy " + policyStr + " but doesn't have the corresponding QcType '" + typeStr + "' (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")"
```

Comment:
I don't see where the requirement to include a QcType exists? The current language being quoted doesn't seem to support this.

This also feels a bit duplicative of the citation. Could you explain a bit more?

Reply:
Modify this implementation to:

if util.HasCertPolicy(c, policy) && util.IsQcStatemPresent(c, &util.IdEtsiQcsQcType) && !util.HasCertEtsiQcType(c, qcType) {

which additionally checks if QcType is set in the certificate. Only then must it be consistent. Ok?

Reply:
I guess what I'm trying to understand is where the requirement is to have a QcType coming from? 

412-4, Clause 4.2 basically just says "Do what clause 4/5 of 412-5" say

412-5, Clause 4.2.3 doesn't mandate the use of QCType, and the interpretation of whether it's an EU Qualified certificate is contingent upon the presence of a QcCompliance as described in Clause 4.2.1. However, even if Clause 4.2.1 is present (QcCompliance), that's still not sufficient, because that depends on whether QcCCLegislation (Clause 4.2.4) is present.

Further, as it pertains to QCP-w's common policy (EN 319 411-2, v2.2.2, Clause 5.5.5), it's unclear whether that is intended to meet the requirements of Annex IV (how 319 412-5 v2.2.3, Table 2 describes it) or whether it's intended to meet articles 3 (38) and 45 of the eIDAS regulation (how 319 411-2, v2.2.2 clause 5.5.5 describes it)

So I think this is very much a separate check from PSD2, but here's my understanding of the check:
- If QcCompliance is asserted, without QcCCLegislation, then:
  - If QcType is present, it's compliant with Annex IV
  - If QcType is absent, it MAY be a Lint violation, or it MAY be a signal that the certificate complies with Directive 1999/99/EC (per Table 1 of 319 412-5).
    - If the QCP-w policy is asserted, then the absence of QcType is a bug, because the QCP-w asserts compliance with eIDAS
    - If there is no QCP-w policy asserted, then it depends on the CA's certification date
- If QcCompliance is asserted, with QcCCLegislation, then:
  - QcType is undefined whether it's required or not, depending on the relevant MOU between the EU and the relevant CCLegislation

But I definitely think teasing this out should be a separate lint, because we have multiple dimensions here to worry about. For example, QcCompliance is clearly required when asserting the policy OIDs from Clause 5.5 of 319 411-2, but it's ambiguous in other cases. The semantics of QcType depend on that assertion, since QcType can be used without QcCompliance to indicate a non-qualified certificate.